### PR TITLE
Rewrite SearchConditionConverter

### PR DESCRIPTION
**Pull request by roji**
Our current SearchConditionConvertingExpressionVisitor has the following design issues:

* It relies on visitor state to track whether the current node is within search condition context (e.g. WHERE) or not.
* Because of this, each and every expression type has to be handled explicitly, e.g. there's a VisitAtTimeZone() which duplicates the generate visitation logic from AtTimeZoneExpression.VisitChildren() - only so that `_isSearchCondition` can be set to false before visiting (and set back at the end).
* This means that every time an expression type is added, SearchConditionConvertingExpressionVisitor must be updated to support it, and duplicates substantial logic (all the expression's VisitChildren()) which must be properly synced/maintained.

This rewrite changes the design as follows:

* No more visitor state; the context (are we in a search condition?) is simply passed via method parameters, and handled generically in Visit().
* Expression types where handling needs to care about the context has methods (e.g. VisitSqlBinary) which accept the context as parameter, passing it down as needed. For all other expression types, regular visitation via VisitChildren is used; this cuts down all the duplicate logic, requiring code only for the very specific expression types we need to care about.
* Various other code cleanups and simplifications.

This is similar to #34697, which changed SqlNullabillityProcessor into a real visitor, allowing VisitChildren to be used (for non-SqlExpression expressions).

/cc @ranma42

### CODE REVIEW

File: src/EFCore.SqlServer/Query/Internal/SearchConditionConverter.cs [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+
+            // A search condition expression in non-search condition context - wrap in CASE/WHEN to convert to bit:
+            // e.g. SELECT a LIKE b => SELECT CASE WHEN a LIKE b THEN 1 ELSE 0 END
+            (false, true) => sqlExpressionFactory.Case(
+                [
+                    new CaseWhenClause(
+                        SimplifyNegatedBinary(sqlExpression),
+                        sqlExpressionFactory.ApplyDefaultTypeMapping(sqlExpressionFactory.Constant(true)))
+                ],
+                sqlExpressionFactory.Constant(false)),
```
Comment (ranma42) [inline src/EFCore.SqlServer/Query/Internal/SearchConditionConverter.cs:86]: as we are re-writing (and explaining) this transformation, we might want to mention that `NULL` is not handled (aka we have a known issue here: https://github.com/dotnet/efcore/issues/34001 )

Comment (roji) [reply src/EFCore.SqlServer/Query/Internal/SearchConditionConverter.cs]: Good idea, done.

File: src/EFCore.SqlServer/Query/Internal/SearchConditionConverter.cs [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+                ? sqlExpressionFactory.Equal(
+                    boolValue
+                        ? sqlExpressionFactory.Constant(1)
+                        : sqlExpressionFactory.Constant(0),
+                    sqlExpressionFactory.Constant(1))
+                : sqlExpressionFactory.Equal(sqlExpression, sqlExpressionFactory.Constant(true)),
+
+            // A search condition expression in non-search condition context - wrap in CASE/WHEN to convert to bit:
+            // e.g. SELECT a LIKE b => SELECT CASE WHEN a LIKE b THEN 1 ELSE 0 END
+            // TODO: NULL is handled properly here, see #34001
```
Comment (ranma42) [inline src/EFCore.SqlServer/Query/Internal/SearchConditionConverter.cs:80]:
```suggestion
            // TODO: NULL is not handled properly here, see #34001
```
😅

Comment (roji) [reply src/EFCore.SqlServer/Query/Internal/SearchConditionConverter.cs]: Thanks... 😅

**Review by maumar — APPROVED** [review]

**Pull request merged**